File: tools/skill-auditor/reporter/analysis.go
package reporter

import (
	"fmt"
	"strings"

	"github.com/pantheon-org/tekhne/tools/skill-auditor/scorer"
)

// Analysis returns a markdown summary of a Result suitable for writing to Analysis.md.
func Analysis(r *scorer.Result) string {
	var sb strings.Builder

	fmt.Fprintf(&sb, "# Skill Audit — %s\n\n", r.Skill)
	fmt.Fprintf(&sb, "**Grade:** %s (%d/%d)\n\n", r.Grade, r.Total, r.MaxTotal)

	fmt.Fprintf(&sb, "## Dimension Scores\n\n")
	fmt.Fprintf(&sb, "| Dimension | Score | Max |\n")
	fmt.Fprintf(&sb, "|---|---|---|\n")
	for _, d := range dimensionOrder {
		score, ok := r.Dimensions[d.key]
		if !ok {
			continue
		}
		fmt.Fprintf(&sb, "| %s | %d | %d |\n", d.label, score, d.max)
	}

	if len(r.ErrorDetails) > 0 || len(r.WarningDetails) > 0 {
		fmt.Fprintf(&sb, "\n## Diagnostics\n")
		if len(r.ErrorDetails) > 0 {
			fmt.Fprintf(&sb, "\n### Errors\n\n")
			for _, d := range r.ErrorDetails {
				fmt.Fprintf(&sb, "- **%s** %s\n", d.Dimension, d.Message)
			}
		}
		if len(r.WarningDetails) > 0 {
			fmt.Fprintf(&sb, "\n### Warnings\n\n")
			for _, d := range r.WarningDetails {
				fmt.Fprintf(&sb, "- **%s** %s\n", d.Dimension, d.Message)
			}
		}
	} else {
		fmt.Fprintf(&sb, "\n## Diagnostics\n\nNo errors or warnings.\n")
	}

	return sb.String()
}

File: crates/skill-auditor/src/reporter/analysis.rs
//! Markdown analysis summary (Go `reporter.Analysis`). Output strings are
//! reproduced verbatim from the Go source for downstream parity, including the
//! punctuation in the headings.

use super::DIMENSION_ORDER;
use crate::scorer::Result;
use std::fmt::Write;

/// Return a markdown summary of a [`Result`] suitable for `Analysis.md`.
pub fn analysis(r: &Result) -> String {
    let mut sb = String::new();

    let _ = write!(sb, "# Skill Audit \u{2014} {}\n\n", r.skill);
    let _ = write!(
        sb,
        "**Grade:** {} ({}/{})\n\n",
        r.grade, r.total, r.max_total
    );

    let _ = write!(sb, "## Dimension Scores\n\n");
    let _ = writeln!(sb, "| Dimension | Score | Max |");
    let _ = writeln!(sb, "|---|---|---|");
    for (key, label, max) in DIMENSION_ORDER {
        let Some(score) = r.dimensions.get(*key) else {
            continue;
        };
        let _ = writeln!(sb, "| {label} | {score} | {max} |");
    }

    if !r.error_details.is_empty() || !r.warning_details.is_empty() {
        let _ = writeln!(sb, "\n## Diagnostics");
        if !r.error_details.is_empty() {
            let _ = write!(sb, "\n### Errors\n\n");
            for d in &r.error_details {
                let _ = writeln!(sb, "- **{}** {}", d.dimension, d.message);
            }
        }
        if !r.warning_details.is_empty() {
            let _ = write!(sb, "\n### Warnings\n\n");
            for d in &r.warning_details {
                let _ = writeln!(sb, "- **{}** {}", d.dimension, d.message);
            }
        }
    } else {
        let _ = writeln!(sb, "\n## Diagnostics\n\nNo errors or warnings.");
    }

    sb
}
